R/statistic-histogram.R

```diff
@@ -138,15 +136,24 @@ histogram.formatRelease <- function(release, n) {
 #' @param h Histogram
 #' @param n Sample size
 
-histogram.compose <- function(h, n) {
-    h <- as.numeric(h)
-    j <- length(h)
-    suppress.index <- sample(1:j, size=1)
-    h[suppress.index] <- 0
-    h[h < 0] <- 0
-    h <- round(h)
-    h[suppress.index] <- n - sum(h)
-    return(h)
+normalizeHistogram <- function(h, n) {
+    # get just the bins sizes for each bin in the histogram (without bin labels)
+    binSizes <- as.numeric(h)
+    
+    # round all negative bins to 0 and all bins greater than n to n
+    binSizes[binSizes < 0] <- 0
+    binSizes[binSizes > n] <- n
+    
+    # get the total size of the histogram, it is not necessarily equal to n
+    sumOfBins <- sum(binSizes)
+    
+    # get a vector of the bin weights
+    binWeights <- binSizes / sumOfBins
+    
+    # get a vector of the normalized bin counts
+    normalizedBinCounts <- binWeights * n
+    
+    return(normalizedBinCounts)
 }
 
 #' Histogram Herfindahl Index
```

Comment:
Add description of Herfindahl index. E.g.

#' Produce differentially private Herfindahl index for categorical types of data.
#' 
#' Let \eqn{s_i} be the percentage of data points in category \eqn{i} in the result histogram. 
#' Then, the Herfindahl index is defined to be the sum of the squares of \eqn{s_i} over all
#' categories. Since the percentage of data points in category %i% may be computed
#' directly from the bin counts in a histogram, a differentially private Herfindahl index 
#' may be calculated from a noisy histogram with no additional privacy loss incurred.